expr/func_aspercent.go

```diff
@@ -0,0 +1,254 @@
+			if _, ok := totalSeries[key]; !ok {
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,MISSING)", serie1.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,MISSING)", serie1.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = math.NaN()
+				}
+			} else {
+				serie2 := totalSeries[key]
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+				}
+			}
+			outSeries = append(outSeries, serie1)
+		}
+	}
+	return outSeries, nil
+}
+
+func (s *FuncAsPercent) execWithoutNodes(series, totals []models.Series) ([]models.Series, error) {
+	var outSeries []models.Series
+	var totalsSerie models.Series
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalsSerie = sumSeries(series)
+		if len(series) == 1 {
+			totalsSerie.Target = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.QueryPatt = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.Tags = map[string]string{"name": totalsSerie.Target}
+		}
+	} else if totals != nil {
+		if len(totals) == 1 {
+			totalsSerie = totals[0]
+		} else if len(totals) == len(series) {
+			// Sorted to match the input series with the total series based on Target.
+			// Mimicks Graphite's implementation
+			sort.Slice(series, func(i, j int) bool {
+				return series[i].Target < series[j].Target
+			})
+			sort.Slice(totals, func(i, j int) bool {
+				return totals[i].Target < totals[j].Target
+			})
+			for i, serie1 := range series {
+				serie2 := totals[i]
+				serie1 = serie1.Copy(pointSlicePool.Get().([]schema.Point))
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+				}
+				outSeries = append(outSeries, serie1)
+			}
+			return outSeries, nil
+		}
+	} else {
+		totalsSerie.QueryPatt = fmt.Sprint(s.totalFloat)
+		totalsSerie.Target = fmt.Sprint(s.totalFloat)
+	}
+
+	for _, serie := range series {
+		serie = serie.Copy(pointSlicePool.Get().([]schema.Point))
+		serie.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie.QueryPatt, totalsSerie.QueryPatt)
+		serie.Target = fmt.Sprintf("asPercent(%s,%s)", serie.Target, totalsSerie.Target)
+		serie.Tags = map[string]string{"name": serie.Target}
+		for i := range serie.Datapoints {
+			var totalVal float64
+			if len(totalsSerie.Datapoints) > i {
+				totalVal = totalsSerie.Datapoints[i].Val
```

Comment:
we may want to always just assume this branch is taken and just write this code line directly without the if/else.
that way if we ever have a bug where `len(totalSerie.Datapoints) != len(serie.Datapoints)` we can troubleshoot it instead of trying to hide such error case and returning incorrect data.

Reply:
this is more of a check of whether there is a totalSeries at all. I'll change it to `> 0`, that way we get an index out of range exception if there's a bug.